src/transformers/models/bloom/modeling_bloom.py

```diff
@@ -0,0 +1,977 @@
+
+    # Loop over the batch where the padding is and replace the alibi tensor by the reference tensor
+    # Only where you do not have padding. Replace padding tokens by zeros
+    # This operation can be seen as a shifting operation.
+    for i, index in enumerate(torch.unique(index_x0)):
+        slice_to_modify = torch.zeros_like(slice_reference_alibi)
+        index_shift = index_y1[index_x1 == index]
+        shift_value = len(index_shift)
+        slice_to_modify[:, :, index_shift] = slice_reference_alibi[:, :, :shift_value]
+        new_alibi[index * num_heads : (index + 1) * num_heads] = slice_to_modify
+    return new_alibi
+
+
+def bias_dropout_add(x, bias, residual, prob, training):
+    # type: (Tensor, Tensor, Tensor, float, bool) -> Tensor
+    out = torch.nn.functional.dropout(x + bias, p=prob, training=training)
+    out = residual + out
+    return out
+
+
+def get_bias_dropout_add(training):
+    def _bias_dropout_add(x, bias, residual, prob):
+        return bias_dropout_add(x, bias, residual, prob, training)
+
+    return _bias_dropout_add
+
+
+@torch.jit.script
+def bias_dropout_add_fused_train(x, bias, residual, prob):
+    # type: (Tensor, Tensor, Tensor, float) -> Tensor
+    return bias_dropout_add(x, bias, residual, prob, True)
+
+
+@torch.jit.script
+def bias_dropout_add_fused_inference(x, bias, residual, prob):
+    # type: (Tensor, Tensor, Tensor, float) -> Tensor
+    return bias_dropout_add(x, bias, residual, prob, False)
+
+
+class BloomScaledSoftmax(nn.Module):
+    """
+    fused operation: scaling + mask + softmax
+
+    Args:
+        input_in_fp16 (`bool`, *required*):
+            flag to indicate if input in fp16 data format.
+        input_in_bf16 (`bool`, *required*):
+            flag to indicate if input in bf16 data format.
+        scaled_masked_softmax_fusion (`bool`, *required*):
+            flag to indicate user want to use softmax fusion
+        mask_func (`function`, *required*):
+            mask function to be applied.
+        softmax_in_fp32 (`bool`, *required*):
+            if true, softmax in performed at fp32 precision.
+        scale (`float`, *required*):
+            scaling factor used in input tensor scaling.
+    """
+
+    def __init__(
+        self,
+        scaled_masked_softmax_fusion,
+        mask_func,
+        softmax_in_fp32,
+        scale,
+    ):
+        super().__init__()
+        self.scaled_masked_softmax_fusion = scaled_masked_softmax_fusion
+        self.mask_func = mask_func
+        self.softmax_in_fp32 = softmax_in_fp32
+        self.scale = scale
+
+        if not (self.scale is None or softmax_in_fp32):
+            raise ValueError("softmax should be in fp32 when scaled")
+
+    def forward(self, input, mask, max_positions):
+        input_dtype = input.dtype
+        input_in_float16 = (input_dtype == torch.float16) or (input_dtype == torch.bfloat16)
+
+        if input_in_float16 and self.softmax_in_fp32:
+            input = input.float()
```

Comment:
I think the same is being discussed in the OPT PR - this is very inefficient!

Reply:
Hmm I would keep it for now since it is done exactly this way in Megatron. Also you can control this parameter in the config file by disabling `attention_softmax_in_fp32`

Reply:
In what sense inefficient? Because it copies the weights for a fp32 precision? That seems to be needed if you want to get a fp32 attention matrix ... Megatron-DeepSpeed does this in it's torch version of `FusedScaleMaskSoftmax` https://github.com/bigscience-workshop/Megatron-DeepSpeed/blob/908dc9cb4b9717707241eaf8b92a986b2e251218/megatron/model/fused_softmax.py#L204-L219 So unless we provide the custom kernel I don't think there's much we can do here.

Reply:
Please see how it ideally should be done here:
https://github.com/huggingface/transformers/pull/17437#issuecomment-1139723683

Reply:
I should have corrected that in e768ab6010a05606e630594b7e208d5a2eb9826f - resolving this conversation!

Reply:
I don't see how this solution is more efficient. `softmax(dtype=torch.float)` also cast the float so the operations are similar (https://pytorch.org/docs/stable/generated/torch.nn.functional.softmax.html). Moreover, this changes the semantics where scaling if done in 16bit instead of 32bit.

Reply:
perhaps the point of this discussion didn't come across in the first place? it's not about similarity but efficiency.

the `dtype` arg isn't just "convert this back to float" it's about doing the accumulation in fp32.

The inputs are already in bf16 and upcasting them to fp32 doesn't change anything as long as any add and multiply operations are accumulated in fp32.

in other words:

`bf16_1.to(fp32)+bf16_2.to(fp32)`

should produce an identical output to:

`accumulate_in_fp32(bf16_1+bf16_2)`

same with multiply.

perhaps the simplest approach here is to write a small data test and confirm that softmax produces the same result as with the manual casting? If it doesn't then perhaps there is a problem with `softmax` as the dtype arg was specifically added to make this situation more efficient.

Reply:
Considering this discussion as resolved!

Reply:
If it doesn't overflow (and it shouldn't in bf16) and there is no concern with lost precision then there is no need. 

Probably best to ask the Megatron-LM team on slack?

Reply:
You mean adding `torch.cuda.synchronized()` on the forward pass of the softmax class would be necessary?

Reply:
(I didn't saw that on Meg-DS script that is why I did not put it)

Reply:
From my experience it's needed to get measure computation time. @stas00 can confirm?

Reply:
why do you want to use `torch.cuda.synchronize()`? 

could we please start a new thread as this is very confusing when topics are switched in the same thread.

Reply:
Going to use the internal slack we have: https://huggingface.slack.com/archives/C01NHER1JLS/p1654039621570599?thread_ts=1654017734.598469&cid=C01NHER1JLS

Reply:
oh, from your comment on slack I got the context of `torch.cuda.synchronize()` you were referring to - yes, this needs to be called between individual groups tests but not during the test itself. i.e. if you compare A to B you sync before A and before B.

Reply:
Yeah sorry for not being explicit, I was mentioning https://github.com/huggingface/transformers/pull/17202#discussion_r885419777 . Effectively I'm scared that the generated benchmark did not return values that were trustworthy.

Reply:
@younesbelkada, I think you didn't paste all of the code in your benchmark - it's missing the reporting code.

I used `timeit` so it does all the work for me and adjusted it to the following, trying to emulate the more realistic use:

```
import torch

n_cycles, n_repeats = 100, 100

input_in_bf_16 = torch.randn(1, 1, 512, 512, dtype=torch.bfloat16).cuda()
#input_in_bf_16 = torch.randn(64, 16, 7, 7, dtype=torch.bfloat16).cuda()
softmax = torch.nn.Softmax(dim=-1)

def explicit_cast1():
    for _ in range(n_cycles): _ = softmax(input_in_bf_16.float())
    torch.cuda.synchronize()

def explicit_cast2():
    for _ in range(n_cycles): _ = torch.nn.functional.softmax(input_in_bf_16.float(), dim=-1)
    torch.cuda.synchronize()

def accumulate():
    for _ in range(n_cycles): _ = torch.nn.functional.softmax(input_in_bf_16, dim=-1, dtype=torch.float32)
    torch.cuda.synchronize()

# 1. timeit
import timeit
print(f'explicit_cast1={timeit.Timer("explicit_cast1()", globals=globals()).timeit(number=n_repeats)}')
print(f'explicit_cast2={timeit.Timer("explicit_cast2()", globals=globals()).timeit(number=n_repeats)}')
print(f'accumulate={timeit.Timer("accumulate()", globals=globals()).timeit(number=n_repeats)}')

# 2. cProfile
import cProfile
cProfile.run("explicit_cast2()", sort=-1)
cProfile.run("accumulate()", sort=-1)
```

I'm actually getting pretty identical results for both functional softmax versions.

```
explicit_cast1=0.2815104750916362
explicit_cast2=0.2630390810081735
accumulate=0.26106692804023623
```

But if you look at the cprofile output you can see that the benchmark is not very clear, since the C kernel isn't being shown - probably `builtins.exec` incorporates it.

So really here we need to use a pytorch profiler instead. 

Probably the easiest version to use is the one with TB-backend https://pytorch.org/tutorials/intermediate/tensorboard_profiler_tutorial.html

output of cprofile
```
   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    0.003    0.003 <string>:1(<module>)
        4    0.000    0.000    0.000    0.000 __init__.py:148(is_initialized)
        4    0.000    0.000    0.000    0.000 __init__.py:190(_lazy_init)
        1    0.000    0.000    0.000    0.000 __init__.py:267(__init__)
        1    0.000    0.000    0.000    0.000 __init__.py:271(__enter__)
        1    0.000    0.000    0.000    0.000 __init__.py:280(__exit__)
        2    0.000    0.000    0.000    0.000 __init__.py:479(current_device)
        1    0.000    0.000    0.000    0.000 __init__.py:485(synchronize)
        1    0.000    0.000    0.000    0.000 __init__.py:76(is_available)
        3    0.000    0.000    0.000    0.000 _jit_internal.py:957(is_scripting)
        1    0.000    0.000    0.000    0.000 _utils.py:460(_get_available_device_type)
        1    0.000    0.000    0.000    0.000 _utils.py:467(_get_device_attr)
        1    0.000    0.000    0.000    0.000 _utils.py:475(_get_current_device_index)
        1    0.000    0.000    0.000    0.000 _utils.py:477(<lambda>)
        1    0.000    0.000    0.000    0.000 _utils.py:499(_get_device_index)
        1    0.000    0.000    0.000    0.000 _utils.py:7(_get_device_index)
        1    0.000    0.000    0.003    0.003 benchmark.py:13(explicit_cast2)
      100    0.000    0.000    0.001    0.000 functional.py:1788(softmax)
        1    0.000    0.000    0.003    0.003 {built-in method builtins.exec}
        1    0.000    0.000    0.000    0.000 {built-in method builtins.hasattr}
        6    0.000    0.000    0.000    0.000 {built-in method builtins.isinstance}
        1    0.000    0.000    0.000    0.000 {built-in method torch._C._cuda_getDeviceCount}
        2    0.000    0.000    0.000    0.000 {built-in method torch._C._cuda_getDevice}
        4    0.000    0.000    0.000    0.000 {built-in method torch._C._cuda_isInBadFork}
        1    0.000    0.000    0.000    0.000 {built-in method torch._C._cuda_synchronize}
      100    0.000    0.000    0.000    0.000 {built-in method torch._C._has_torch_function_unary}
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}
      100    0.001    0.000    0.001    0.000 {method 'float' of 'torch._C._TensorBase' objects}
        1    0.000    0.000    0.000    0.000 {method 'lower' of 'str' objects}
      100    0.001    0.000    0.001    0.000 {method 'softmax' of 'torch._C._TensorBase' objects}


         344 function calls in 0.003 seconds

   Ordered by: standard name

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    0.003    0.003 <string>:1(<module>)
        4    0.000    0.000    0.000    0.000 __init__.py:148(is_initialized)
        4    0.000    0.000    0.000    0.000 __init__.py:190(_lazy_init)
        1    0.000    0.000    0.000    0.000 __init__.py:267(__init__)
        1    0.000    0.000    0.000    0.000 __init__.py:271(__enter__)
        1    0.000    0.000    0.000    0.000 __init__.py:280(__exit__)
        2    0.000    0.000    0.000    0.000 __init__.py:479(current_device)
        1    0.000    0.000    0.000    0.000 __init__.py:485(synchronize)
        1    0.000    0.000    0.000    0.000 __init__.py:76(is_available)
        3    0.000    0.000    0.000    0.000 _jit_internal.py:957(is_scripting)
        1    0.000    0.000    0.000    0.000 _utils.py:460(_get_available_device_type)
        1    0.000    0.000    0.000    0.000 _utils.py:467(_get_device_attr)
        1    0.000    0.000    0.000    0.000 _utils.py:475(_get_current_device_index)
        1    0.000    0.000    0.000    0.000 _utils.py:477(<lambda>)
        1    0.000    0.000    0.000    0.000 _utils.py:499(_get_device_index)
        1    0.000    0.000    0.000    0.000 _utils.py:7(_get_device_index)
        1    0.000    0.000    0.003    0.003 benchmark.py:17(accumulate)
      100    0.000    0.000    0.003    0.000 functional.py:1788(softmax)
        1    0.000    0.000    0.003    0.003 {built-in method builtins.exec}
        1    0.000    0.000    0.000    0.000 {built-in method builtins.hasattr}
        6    0.000    0.000    0.000    0.000 {built-in method builtins.isinstance}
        1    0.000    0.000    0.000    0.000 {built-in method torch._C._cuda_getDeviceCount}
        2    0.000    0.000    0.000    0.000 {built-in method torch._C._cuda_getDevice}
        4    0.000    0.000    0.000    0.000 {built-in method torch._C._cuda_isInBadFork}
        1    0.000    0.000    0.000    0.000 {built-in method torch._C._cuda_synchronize}
      100    0.000    0.000    0.000    0.000 {built-in method torch._C._has_torch_function_unary}
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}
        1    0.000    0.000    0.000    0.000 {method 'lower' of 'str' objects}
      100    0.003    0.000    0.003    0.000 {method 'softmax' of 'torch._C._TensorBase' objects}
```